pkg/geodata/geodata.go

```diff
@@ -414,68 +447,111 @@ func censusTableFromAndSQL(censustable string) (string, string) {
-		namedCatSQL = ""
+	namedCatSQL := where.WherePart("nomis_category.long_nomis_code", set)
+	if namedCatSQL != "" {
+		conditions = append(conditions, namedCatSQL)
 	}
-	// get sql for selecting categories by nomis desc selection - will need joining 'OR' if there were named cats
-	var censusTableSQL string
+
+	// get sql for selecting categories by nomis desc selection
 	if censusTable != "" {
-		censusTableSQL = " nomis_category.nomis_desc_id = nomis_desc.id"
-		if namedCatSQL != "" {
-			censusTableSQL = " OR\n " + censusTableSQL
-		}
+		conditions = append(conditions, " nomis_category.nomis_desc_id = nomis_desc.id")
 	}
+
+	if len(conditions) == 0 {
+		return "", nil
+	}
+
 	template := `
 AND (
 %s
-%s
 )`
-	return fmt.Sprintf(template, namedCatSQL, censusTableSQL), nil
+	return fmt.Sprintf(template, strings.Join(conditions, " OR\n ")), nil
 }
 
-// splitCols separates special column names from geography names
-// (XXXX This duplicates some of the work in where.GetValues().
-// Think of a better way.)
-func splitCols(cols []string) (include []string, cats []string) {
-	for _, instance := range cols {
-		tokens := strings.Split(instance, ",")
-		for _, token := range tokens {
-			if token == table.ColGeographyCode || token == table.ColGeotype {
-				include = append(include, token)
-			} else {
-				cats = append(cats, token)
+// ExtractSpecialCols removes special column names like "geography_code" from
+// the ValueSet, returning a reduced ValueSet, and the list of special columns
+// found.
+func ExtractSpecialCols(set *where.ValueSet) ([]string, *where.ValueSet, error) {
+	var includes []string
+
+	callback := func(single, low, high *string) (*string, *string, *string, error) {
+		var err error
+		if single != nil {
+			if *single == table.ColGeographyCode || *single == table.ColGeotype {
+				includes = append(includes, *single)
+				single = nil
 			}
+		} else {
+			if *low == table.ColGeographyCode || *high == table.ColGeographyCode || *low == table.ColGeotype || *high == table.ColGeotype {
+				err = fmt.Errorf("%w: special columns cannot be part of a range", ErrInvalidParams)
+			}
+		}
+		if err != nil {
+			return nil, nil, nil, err
 		}
+		return single, low, high, nil
 	}
-	return
+
+	newset, err := set.Walk(callback)
+	return includes, newset, err
 }
 
 // additionalCondition wraps the output of WherePart inside "AND (...)".
 // We "know" this additionalCondition will not be the first additionalCondition in the query.
 func additionalCondition(col string, args []string) (string, error) {
-	if len(args) == 0 {
-		return "", nil
+	set, err := where.ParseMultiArgs(args)
+	if err != nil {
+		return "", err
 	}
-	body, err := where.WherePart(col, args)
+	set, err = MapGeotypes(set)
 	if err != nil {
 		return "", err
 	}
+	body := where.WherePart(col, set)
+
+	if body == "" {
+		return "", nil
+	}
 
 	template := `
 AND (
 %s
 )`
 	return fmt.Sprintf(template, body), nil
 }
+
+func FixGeotype(token string) (string, error) {
```

Comment:
From the function name I'd guess this attempts to coerce the geotype token into a valid value, but it just errors if its invalid, right?